README.md
# Deprecation

:rotating_light: **Assets Frontend has been deprecated in favour of the [Play Frontend](https://www.github.com/hmrc/play-frontend-hmrc).** :rotating_light:

If you do decide to use Assets Frontend anyway you need to be aware of the vulnerabilities including but not limited to the ones listed here ]https://github.com/hmrc/assets-frontend/security/dependabot](https://github.com/hmrc/assets-frontend/security/dependabot).  There are also known accessibility issues when using Assets Frontend which will cause delays with going live.  We won't be making updates to Assets Frontend so we recommend moving over to the latest [Play Frontend Libraries](https://confluence.tools.tax.service.gov.uk/display/PLATUI/play-frontend+Libraries)

# assets-frontend

This repository contains the source files and documentation for the Components & Design Patterns used to design and build digital services for the HMRC Tax Platform.

It provides additional styles on top of the [GOV.UK styles](govuk-elements).


# Quick Start

## Requirements

* [Node.js](https://nodejs.org/en/) `== 4.8.4`
* [npm](https://www.npmjs.com/) `>= 2.11.3`

To install multiple versions of Node.js, you may find it easier to use a node version manager:

* [nvm](https://github.com/creationix/nvm)
* [n](https://github.com/tj/n)

## Installation

Clone this repository and install its dependencies:

```
$ git clone https://github.com/hmrc/assets-frontend.git
$ cd assets-frontend
$ npm install
```

## Running

### The Design System:

1. Run `npm start`
2. The Design System is available at http://localhost:9034/
3. Latest V4 Assets are then available at http://localhost:9032/assets/v4-SNAPSHOT/...
4. Latest V3 Assets are then available at http://localhost:9032/assets/v3-SNAPSHOT/...

### To access both the Design System and the Component Library [DEPRECATED]

1. Run `npm run dev:all` or `./server.sh dev`
2. The Design System is available at http://localhost:9034/
3. The Component Library is available at http://localhost:9033/

### To run a specific version of Assets Frontend locally

1. Run `./server.sh`
2. V4 Assets are then available at http://localhost:9032/assets/4.11.0/...
3. V3 Assets are then available at http://localhost:9032/assets/3.11.0/...

# Usage

## Using assets locally

### Prototypes

For now, the quickest and simplest way to add assets-frontend to your prototype is to grab the built CSS and JS from production.

Just replace `<VERSION>` in the links below with the [released version](https://github.com/hmrc/assets-frontend/releases) you want to use (we recommend using the latest).

If your prototype is based on the [GOV.UK prototype kit](https://github.com/alphagov/govuk_prototype_kit/) then do the following:

**CSS**

Save the file found at
[https://www.tax.service.gov.uk/assets/\<VERSION\>/stylesheets/application.min.css](https://www.tax.service.gov.uk/assets/\<VERSION\>/stylesheets/application.min.css)
to `/app/assets/sass/assets-frontend.css`.

And add the following to `app/views/includes/head.html`:
```
<link href="/public/stylesheets/assets-frontend.css" rel="stylesheet" type="text/css" />
```

**JavaScript**

Save the file found at
[https://www.tax.service.gov.uk/assets/\<VERSION\>/javascripts/application.min.js](https://www.tax.service.gov.uk/assets/\<VERSION\>/javascripts/application.min.js)
to `/app/assets/javascripts/assets-frontend.js`

And add the following to `app/views/includes/head.html`
```
<script src="/public/javascripts/assets-frontend.js"></script>
```

## Frontends

If your frontend is based on [init-service](https://github.com/hmrc/init-service/), then you just need to make sure the configuration for assets in the [application.conf](https://github.com/hmrc/init-service/blob/f9a55c100faa8b13d2a1a869c0531f6e3a7b556c/templates/service/conf/application.conf#L73-L77) file has the line `version = ${?ASSETS_FRONTEND_VERSION}` after declaring the actual version.

Once it has, you just need to set an environment variable before running your frontend:

```
$ ASSETS_FRONTEND_VERSION=999-SNAPSHOT sbt run
```

If you’re using [service-manager](https://github.com/hmrc/service-manager), please read the [service manager guidance in the Wiki](https://github.com/hmrc/assets-frontend/wiki/Using-service-manager).


## Using assets in production

Running `npm run release` calls the [release gulp task](https://github.com/hmrc/assets-frontend/blob/main/gulpfile.js/tasks/release.js) and this happens on our internal Jenkins instances.

The final step of our internal release pipeline is to version and deploy the compiled assets to https://www.tax.service.gov.uk/assets/.

Releasing assets for use in production is currently a process owned by the [Service Design Tools](https://github.com/orgs/hmrc/teams/service-design-tools) team.

If you’d like a new version of assets-frontend released, please get in touch with us in the [#team-sdt](https://hmrcdigital.slack.com/messages/C39V3PH38) Slack channel.


# HMRC Design System

The [HMRC Design System can be viewed here](http://hmrc.github.io/assets-frontend/).

For detailed information on how the HMRC Design System works, please read the [Design System page in the Wiki](https://github.com/hmrc/assets-frontend/wiki/HMRC-Design-System).


# Component Library - [DEPRECATED]

:rotating_light: **The Component Library has been deprecated in favour of the [HMRC Design System](http://hmrc.github.io/assets-frontend/).**:rotating_light:

If your service relies on a component or pattern that doesn’t appear in the HMRC Design System then please follow the [Design System contribution guidelines](https://github.com/hmrc/assets-frontend/wiki/HMRC-Design-System#contributing-a-design-pattern).

For detailed information on how the Component Library works, please read the [Component Library guidance in the Wiki](https://github.com/hmrc/assets-frontend/wiki/Component-Library-%5BDEPRECATED%5D).


# Contributing

## Design Patterns

For details on how to contribute Design Patterns, please take a few minutes to review our [Design Pattern standards and contribution process](https://github.com/hmrc/assets-frontend/wiki/HMRC-Design-System#contributing-a-design-pattern) before you submit your request, otherwise it may be rejected.

## Features and issues

If you’ve spotted an issue or thought of a feature that you’d like to contribute to assets-frontend, please take a few minutes to review our [contribution process and guidelines for Assets Frontend](CONTRIBUTING.md) before you submit your request, otherwise it may be rejected.


# License

This code is open source software licensed under the [Apache 2.0 License]("http://www.apache.org/licenses/LICENSE-2.0.html").


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [4.22.0] and [3.22.0] - 2020-08-19
### Changed
- Bumped dependencies to resolve vulnerabilities

## [4.21.0] and [3.21.0] - 2020-07-02
### Changed
- Bumped dependencies to resolve vulnerabilities
- Using specific versions for precision
- Not using mkdirp as a dependency as it was less work to rebuild it now that recursive directories are supported directly in fs
- Updated message for analytics stub

## [4.20.0] and [3.20.0] - 2020-07-01
### Changed
- Removed stageprompt as it delivers Google Analytics in a way which shouldn't be used alongside 
  [hmrc/tracking-consent-frontend](https://www.github.com/hmrc/tracking-consent-frontend)


## [4.19.0] and [3.19.0] - 2020-06-15
### Changed
- Removed old cookie banner functionality replaced by 
  [hmrc/tracking-consent-frontend](https://www.github.com/hmrc/tracking-consent-frontend)

## [4.18.0] and [3.18.0] - 2020-02-22
### Security
- Patch security vulnerabilities

## [4.17.0] and [3.17.0] - 2020-01-28
### Changed
- Inlined hmrc/node-git-versioning

## [3.16.1] and [4.16.1] - 2020-12-11
### Security
- Bump version of ini to 1.3.7

## [3.15.0] and [4.15.0] - 2020-10-30
### Fixed
- Uplifted jQuery version to 1.12.4, and added patch to assets/application.js, to fix the vulnerability [CVE-2020-11022](http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-11022)
- Uplifted browser-sync version to 2.26.13 to fix the vulnerability in the dependent package object-path [CVE-2020-15256](http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15256)

## [3.14.0] and [4.14.0] - 2020-10-20
### Fixed
- Documented the archiving of component-library-template
- Incorrectly added aria-hidden attribute on 'Your account' link on PTA account menu
- Added WCAG 2.1 AA focus and colour contrast fixes from
https://gist.github.com/adamliptrot-oc/f77250a6f69fb31fabd935e2002f4964 (v3)
and https://gist.github.com/adamliptrot-oc/c35166dce1f3234bfc8f0dbb915f07b8 (v4)

### Updated
- Documented the archiving of component-library-template

## [3.13.0] and [4.13.0] - 2020-09-15
### Fixed
- Fixed Puppeteer configuration and added --no-sandbox
- Node uplifted to v12 for consistency with hmrc-frontend
- Gulp uplifted to v4 and pipelines refactored
- kss uplifted and component-library custom builder re-built
- PhantomJS replaced with Puppeteer
- Elements from hmrc/component-library-template integrated into assets-frontend
- VRT tests re-instated
- Various security advisory reports resolved (handlebars, elliptic, bl,
marked, dot-prop, node-sass, jpeg-js, lodash) 
- Unused dependencies removed

## [3.12.0] and [4.12.0] - 2020-07-31
### Fixed
- Improved accessibility of timeoutDialog.
- Dependencies with known vulnerabilities have been updated.
- Fixed the leak detection issue for the repository.

## [3.11.0] and [4.11.0] - 2019-04-11
### Fixed
- When navigating through the navigation bar, 'your account' doesn't highlight when the focus is on it. To fix this, CSS removed that was overriding default behavious of highlighting on focus.


## [3.10.0] and [4.10.0] - 2019-04-03
### Fixed
- The assets path in the static error pages was pointing to a SNAPSHOT version of the assests since new build changes
- The javascript link in the static error pages was pointing to an unminified filename which isn't present in the release

## [3.9.0] and [4.9.0] - 2019-01-24
### Updated
- Updated VAT registration number documentation [#1012](https://github.com/hmrc/assets-frontend/pull/1012)
- Added a deprecation banner to direct users to the new design resources for HMRC on Heroku, with associated content and styling changes [#1011](https://github.com/hmrc/assets-frontend/pull/1011)
- Added an LDS exception for the design-system that will be copied across to the gh-pages branch [#1013](https://github.com/hmrc/assets-frontend/pull/1012)

### Added
- Optional full width layout for tables [#1014](https://github.com/hmrc/assets-frontend/pull/1014)

## [3.8.0] and [4.8.0] - 2018-12-14
### New
- Added mitigation for the vulnerability [CVE-2015-9251](http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-9251) in older versions of jQuery [#1009](https://github.com/hmrc/assets-frontend/pull/1009)
- Added expanding subnavigation to mobile navigation [#1004](https://github.com/hmrc/assets-frontend/pull/1004)

### Fixed
- Updated the fixture and tests for the design library template [#1005](https://github.com/hmrc/assets-frontend/pull/1005)

### updated
- Updated documentation following a prototype review [#1008](https://github.com/hmrc/assets-frontend/pull/1008)

### Removed
- Removed unused nodemon package that's affected by [the `flatmap-stream` vulnerability](https://github.com/dominictarr/event-stream/issues/116)[#1007](https://github.com/hmrc/assets-frontend/pull/1007)

## [3.7.0] and [4.7.0] 2018-11-19
### New
- Updated confirmed identity and could not confirm identity patterns [#1000](https://github.com/hmrc/assets-frontend/pull/1000)
- Added references and identifiers to system [#996](https://github.com/hmrc/assets-frontend/pull/996)

### Updated
- Updated add to a list code and animated gif [#1002](https://github.com/hmrc/assets-frontend/pull/1002)
- Updated example on identity page [#998](https://github.com/hmrc/assets-frontend/pull/998)
- Updated identity pages [#997](https://github.com/hmrc/assets-frontend/pull/997)
- Updated service unavailable page example [#995](https://github.com/hmrc/assets-frontend/pull/995)
- Changed accent colour on the [documentation](https://hmrc.github.io/assets-frontend) [#999] (https://github.com/hmrc/assets-frontend/pull/999)

### Fixed
- Fixed validorFocus.js to escape dots in jquery id selector [#984](https://github.com/hmrc/assets-frontend/pull/984)

gulpfile.js/config.js
'use strict'

/*
- Clean all generated dirs
- Copy error pages
- Styles
- Scripts
- Images

Dev:
- Sourcemaps
- BrowserSync
- Watch

Build:
- Image compression

Test:
- Continuous
- Spec reporter
*/
const src = './assets/'
const dest = src + 'public/'
const distDir = src + 'dist/'

var govuk = {
  elements: 'node_modules/govuk-elements-sass',
  toolkit: 'node_modules/govuk_frontend_toolkit'
}

module.exports = {
  src: src,

  dest: dest,

  distDir: distDir,

  designSystemRepositoryMetadata: 'design-system-repository-metadata.yaml',

  snapshotDir: {
    v3: dest + 'v3-SNAPSHOT/',
    v4: dest + 'v4-SNAPSHOT/'
  },

  scripts: {
    src: [
      src + 'javascripts/modules/**/*.js',
      src + 'javascripts/**/index.js',
      src + 'components/**/*.js',
      src + 'application.js',
      '!**/*.polyfill.js'
    ],
    destDirName: 'javascripts',
    vendorDestDirName: 'javascripts/vendor',
    gulpTasks: 'gulpfile.js/**/*.js',
    encryptionSrc: src + 'javascripts/encryption/**/*.js',

    modernizr: {
      options: [
        'setClasses',
        'html5printshiv',
        'testProp'
      ],
      tests: [
        'touchevents'
      ],
      excludeTests: [
        'flash',
        'hidden',
        'arrow'
      ],
      files: {
        src: [
          src + '{javascripts,scss}/**/*.{js,scss}',
          govuk.toolkit + '/**/*.{js,scss}',
          govuk.elements + '/**/*.{js,scss}',
          '!**[^node_modules]/**/modernizr.js'
        ]
      }
    }
  },

  svg: {
    src: [
      src + 'images/**/*.svg',
      govuk.toolkit + '/images/**/*.svg'
    ],
    destDirName: 'images'
  },

  images: {
    src: [
      src + 'images/**/*',
      '!' + src + 'images/**/*.svg',
      govuk.toolkit + '/images/**/*',
      '!' + govuk.toolkit + '/images/**/*.svg'
    ],
    destDirName: 'images'
  },

  sass: {
    src: [
      src + '*.scss',
      src + 'scss/**/*.scss',
      src + 'components/**/*.scss'
    ],
    destDirName: 'stylesheets',
    settings: {
      sourceComments: true,
      includePaths: [
        govuk.elements + '/public/sass',
        govuk.toolkit + '/stylesheets'
      ],
      outputStyle: 'compressed'
    },
    sourceMapsDir: './maps'

  },

  browserify: {
    bundleConfigs: [{
      basedir: './',
      entries: [
        src + 'application.js'
      ],
      add: [
        govuk.toolkit + '/javascripts/govuk/**/*.js',
        '!' + govuk.toolkit + '/javascripts/govuk/selection-buttons.js',
        '!' + govuk.toolkit + '/javascripts/govuk/analytics/*'
      ],
      destDirName: 'javascripts',
      outputName: 'application.js'
    }, {
      entries: [
        src + 'javascripts/export/fingerprint.js'
      ],
      destDirName: 'javascripts',
      outputName: 'mdtpdf.js'
    }]
  },

  test: {
    src: src + 'test/**/*.js',
    files: {
      v3: [
        'test/specs/fixtures/*.html',
        'test/specs/*.js',
        'components/**/*.html',
        'components/**/*.test.js',
        'components/**/*.test-mobile.js',
        'patterns/**/*.html',
        'patterns/**/*.test.js',
        'public/v3-SNAPSHOT/stylesheets/application.min.css'
      ],
      v4: [
        'components/**/*.html',
        'components/**/*.test.js',
        'components/**/*.test-mobile.js',
        'patterns/**/*.html',
        'patterns/**/*.test.js',
        'public/v4-SNAPSHOT/stylesheets/application.min.css'
      ]
    },
    specsScr: src + 'test/specs/unit/**/*.js',
    fixturesScr: src + 'test/specs/fixtures/*.html',
    karmaConfig: src + 'test/config/karma.conf.js',
    gulpTasks: [
      'gulpfile.js/tests/**/*.test.js',
      '!gulpfile.js/tests/fixtures/**/*'
    ]
  },

  errorPages: {
    src: src + 'error_pages/*.html',
    assetsBaseUri: '/assets/'
  },

  componentLibrary: {
    baseDir: './component-library/',
    dest: 'component-library'
  },

  designSystem: {
    sourceBaseDir: src,
    src: [
      src + 'styles',
      src + 'components',
      src + 'patterns'
    ],
    dest: 'design-system',
    template: './gulpfile.js/util/design-system/design-system.html',
    homepage: 'design-system.md',
    helpers: './gulpfile.js/util/design-system/helpers',
    macrosPath: './gulpfile.js/util/design-system/macros'
  },

  browserSync: [{
    ui: false,
    port: 9032,
    open: false,
    logLevel: 'silent',
    host: '127.0.0.1',
    server: {
      baseDir: '.',
      routes: {
        '/assets': dest
      }
    }
  }, {
    ui: false,
    port: 9033,
    open: false,
    logLevel: 'silent',
    localOnly: true,
    host: '127.0.0.1',
    server: 'component-library'
  }, {
    ui: false,
    port: 9034,
    open: false,
    logLevel: 'silent',
    localOnly: true,
    host: '127.0.0.1',
    server: 'design-system'
  }]
}
